New Messagesystem, Un/Register Connector, Reconstruct code

frenzqse opened the pull request.

frenzqse committed:
[OPENENGSB-3026] Change messagestructure
[OPENENGSB-3025] Rename Namespaces

frenzqse committed:
[OPENENGSB-3027] Restructor the code

frenzqse committed:
[OPENENGSB-3028] Add System.Web.Services as references for csc in the wsdlToDll plugin
[OPENENGSB-3026] Change messagestructure (Remove Bug)

frenzqse committed:
[OPENENGSB-2925] Increase log statements in bridge

frenzqse committed:
[OPENENGSB-3038] .Net bridge and OpenEngSBModelWrapper

frenzqse committed:
[OPENENGSB-3039] Add licence, format and organise uses in .Net bridge
Remove Classes for testing the PLC domain

frenzqse committed:
[OPENENGSB-3042]Bug in CreateResultMethod of the .Net bridge

frenzqse committed:
[OPENENGSB-3046] Add a way to check if the .Net Bridge registered completly

frenzqse committed:
[OPENENGSB-3040] Adapte pom of the .Net bridge to the new added/changed classes

frenzqse committed:
[OPENENGSB-3050] Rename RegisterDomainService and UnregisterDomainServic in .Net bridge
[OPENENGSB-3051] Integrate registration and unregistration of a connector in the .Net bridge

frenzqse committed:
[OPENENGSB-3055] Adaptation so that it is possible to build the Net bridge

frenzqse committed:
[OPENENGSB-3059] Add parameter registrationId in the .Net Bridge, to register with the openEngSB

ChristophGr commented on bridge/Implementation/Common/Enumeration/ERegistration.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public enum ERegistration
+    {
+        REGISTERED, CONNECTORCREATED, NONE
```
Either prefix both states with "CONNECTOR" or not.
(I prefer not)

ChristophGr commented on bridge/Implementation/Common/IStoppable.cs:
```diff
@@ -29,5 +23,6 @@ public interface IStoppable
     {
         //Description to stop a Thread
         void Stop();
+        void UnRegisterConnector();
```
IStoppable seems to generic for this method.

ChristophGr replied:
Rather rename them to "Stop()", and "Delete()"

frenzqse replied:
Well stop is for stoping the application. The stop implies a Unregistration

ChristophGr replied:
What does stop do, what unregisterConnector doesn't?
Does Stop() delete the connector?

frenzqse replied:
Stop invokes that the connector gets deleted and actviemq will be closed. Unregister connector just unregisters the connector. The connector is still oben. for example

registerConnector
unregisterConnector
registerConnector
unregisterConnector

is possible but

registerConnector
unregisterConnector
stop();
registerConntror => exception

ChristophGr replied:
Actually I think "stopping" the connector should not delete it.
It should only be deleted when it is not going to be used anymore (ever!).

frenzqse replied:
Well I thing for it and comment it tommorow

ChristophGr commented on bridge/Implementation/HelpMethods.cs:
```diff
@@ -48,25 +50,68 @@ public static String reverseURL(String url)
+                        {
+                            String[] element = s.Split(':');
+                            if (element.Length>1 && element[0].ToUpper().Equals(name.ToUpper())) return element[1];
+                        }
```
This code has many indention levels. Try to extract some methods.

frenzqse replied:
This was a little hack for the cxf plugin. I was thinking to remove it because we would use cxf anymore. Should I remove it or rewrite it?

ChristophGr replied:
If it's dead code, just delete it. If we need it later we have an SCM for that. If not someone will forget why it was there in the first place (see Lava Flow Antipattern).

ChristophGr commented on bridge/Implementation/HelpMethods.cs:
```diff
@@ -48,25 +50,68 @@ public static String reverseURL(String url)
+            foreach (Attribute attribute in type.GetCustomAttributes(false))
+            {
+                if (attribute is WebServiceBindingAttribute)
+                {
```
You could use a guard to prevent this additional indention-level

ChristophGr commented on bridge/Implementation/HelpMethods.cs:
```diff
@@ -48,25 +50,68 @@ public static String reverseURL(String url)
                     {
                         soapAttribute = attribute as SoapDocumentMethodAttribute;
-                        return reverseURL(soapAttribute.RequestNamespace);
+                        result = reverseURL(soapAttribute.RequestNamespace);
```
I liked it better with the return inside the for-loop. Try to extract the other part into a separete method rather than declaring this odd local variable.

frenzqse replied:
Well I prefered the return too but that had to do with the cxf plugin. When I remove this part totally, I can rewrite this to a beautifull way

frenzqse replied:
(To support both wsdls from the plugins at the same time)

ChristophGr commented on bridge/Implementation/HelpMethods.cs:
```diff
@@ -106,12 +154,13 @@ public static String FirstLetterToUpper(String element)
                 String paramName = paraminfo[i].Name + "Specified";
-                if ((paraminfo[i + 1].ParameterType.Equals(typeof(System.Boolean))) && paramName.Equals(paraminfo[i + 1].Name)) args.Insert(i + 1, true);
+                if ((paraminfo[i + 1].ParameterType.Equals(typeof(System.Boolean))) && 
+                    paramName.Equals(paraminfo[i + 1].Name)) args.Insert(i + 1, true);
```
Please do not mix conditions and instructions in the same line.

ChristophGr commented on bridge/Implementation/HelpMethods.cs:
```diff
@@ -106,12 +154,13 @@ public static String FirstLetterToUpper(String element)
-            if (paraminfo.Length <= args.Count) return;
+            if (paraminfo.Length <= args.Count && paraminfo.Length < 2 && args.Count <= 0) return;
             int i = 0;
             while (i + 1 < paraminfo.Length)
```
Can't you use a foreach loop here?

ChristophGr commented on bridge/Implementation/HelpMethods.cs:
```diff
@@ -106,12 +154,13 @@ public static String FirstLetterToUpper(String element)
         {
             ParameterInfo[] paraminfo = m.GetParameters();
-            if (paraminfo.Length <= args.Count) return;
+            if (paraminfo.Length <= args.Count && paraminfo.Length < 2 && args.Count <= 0) return;
```
Why < 2?

frenzqse replied:
This is a hake, wsdl.exe generate for boolean fields always a is specified parameter. This has to be true, else the handling of the method is incorrect. At This means that there has to be at least 2 parameters. For example foobar(boolean a1, boolean a1Specified)

ChristophGr commented on bridge/Implementation/HelpMethods.cs:
```diff
@@ -143,39 +192,39 @@ public static int AddTrueForSpecified(List<ParameterInfo> parameterResult, Metho
         /// <returns></returns>
+        public static bool TypeModuleAreEqual(String args, ParameterInfo[] parameterInfos)
+        {            
+            OpenEngSBModelWrapper element = (OpenEngSBModelWrapper)(new JsonMarshaller()).UnmarshallObject(args, typeof(OpenEngSBModelWrapper));
```
Is it possible to cache the instance of the JsonMarhsaller (in a static variable)?

frenzqse replied:
Yes and it is. I forgot to change this part

frenzqse replied:
THis method can be removed. I replaced it with something else

ChristophGr commented on bridge/Implementation/OpenEngSB2_4_0/RealDomainFactory.cs:
```diff
@@ -131,6 +135,22 @@ public String getDomainTypServiceId()
+        {
+            return serviceId;
+        }
+        public void RegisterConnector<T>(string registerId, string destination, T service, string domainType)
```
OpenEngSB-2.4.0 does not contain the register-methods and never will, because it breaks API

frenzqse replied:
That why it throws a NotImplementedException.

ChristophGr replied:
Ah, it's defined in the Interface, alright.

ChristophGr commented on bridge/Implementation/OpenEngSB2_4_0/Remote/DomainProxy.cs:
```diff
@@ -16,64 +16,29 @@
+        public DomainProxy(string host, string serviceId, String domainType)
+            : base(host, serviceId, domainType)
+        {
+            AUTHENTIFICATION_CLASS = "org.openengsb.core.api.security.model.UsernamePasswordAuthenticationInfo";
```
Rather use an actual "const" for that?

ChristophGr commented on bridge/Implementation/OpenEngSB2_4_0/Remote/DomainReverseProxy.cs:
```diff
@@ -104,21 +41,14 @@ public T DomainService
-            this.username = "admin";
-            this.password = "password";
+            logger.Info("Connecting to OpenEngSB version 2.4");
+            CREATION_METHOD_NAME = "create";
```
use constants?

ChristophGr commented on bridge/Implementation/OpenEngSB2_4_0/Remote/DomainReverseProxy.cs:
```diff
@@ -332,103 +241,13 @@ private MethodResultMessage CreateMethodReturn(MethodResult.ReturnType type, obj
-                portIn.Close();
-                DeleteRemoteProxy();
-            }
+            throw new MissingMethodException("Not implemented for this version of openEngSB");
```
Maybe also state it's not necessary

ChristophGr commented on bridge/Implementation/OpenEngSB3_0_0/Remote/DomainProxy.cs:
```diff
@@ -17,71 +17,29 @@
-            this.marshaller = new JsonMarshaller();
-            this.username = "admin";
-            this.password = "password";
+            AUTHENTIFICATION_CLASS = "org.openengsb.connector.usernamepassword.Password";
```
Const

ChristophGr commented on bridge/Implementation/OpenEngSB3_0_0/Remote/DomainReverseProxy.cs:
```diff
@@ -104,19 +42,12 @@ public T DomainService
-            this.username = "admin";
-            this.password = "password";
+            logger.Info("Connecting to OpenEngSB version 3.0");
+            CREATION_METHOD_NAME = "createWithId";
```
const

frenzqse replied:
Well the thing with const is a bit bad in C#. Const can never be assign (not in the constructor). Then there is readonly (like final in java), which can be set in the constructor but not in subConstructors. Because of the OpenEngSB v2.4 version I have to set these diffrent. (Like createWithId in 3.0 and create in 2.4). This was the nicessest way.

ChristophGr replied:
Then create two separate constants for that.
Like:
CREATION_METHOD_NAME_2_X
CREATION_METHOD_NAME_3_X

frenzqse replied:
but why? When I set the variables in the constructor, then I can reuse alot of code.

ChristophGr replied:
Then create additional variables for that, but these "magic strings" need to be in actual constants.

ChristophGr replied:
And follow the name convention. All-caps names are for constants.

frenzqse replied:
I found a way, how I solve this. I greated a abstract proprety (getter) with has to be overriten in the sub classes. THis proprety is read only, so it is a like a const

ChristophGr commented on bridge/Implementation/OpenEngSB3_0_0/Remote/DomainReverseProxy.cs:
```diff
@@ -127,140 +58,163 @@ public DomainReverseProxy(T localDomainService, string host, string serviceId, s
-            Guid id = Guid.NewGuid();
-
-            String classname = "org.openengsb.connector.usernamepassword.Password";
+            //registerId = Guid.NewGuid().ToString();
```
don't just comment out code, delete it.

ChristophGr commented on bridge/Interface/IDomainFactory.cs:
```diff
@@ -14,20 +14,19 @@
-        String getDomainTypServiceId();        
+        void CreateDomainService<T>(string destination, T service, string domainType);
+        void DeleteDomainService(object service);
+        void RegisterConnector<T>(string registrationid, string destination, T service, string domainType);
```
The "destination" should actually be managed  by the bridge I think.
The domain-type should not be required in this method.

frenzqse replied:
The destinition is the url of the openEngSB. This has be be provided. The domainType is the domainName like Example and PLC. These are required too.

ChristophGr replied:
Ah, in loom-java, I designed this to be an attribute of the DomainFactory.
Why does the DomainFactory contain attributes like "serviceId" and "domainType" but not "destination"?
Is that thing thread-safe?

frenzqse replied:
Well my domainfactory search the correct implementation for the correct openEngSB version. With one factory, you can connect to serveral domains on diffrent location. I saw this as a feature. On the other hand, including the paramets in the factory is also possible.

frenzqse replied:
Btw. Yes it is thread save. It creates for each connector a complet new instance with a new serviceId

ChristophGr commented on bridge/ServiceTestConsole/Program.cs:
```diff
@@ -37,29 +33,31 @@ class Program
             string destination = "tcp://localhost.:6549";
             string domainName = "example";
+            logger.Info("\n------------------------------------------------"
+            + "\n     Start Example wit the domain " + domainName
```
Please don't use such log-highlighting-art

frenzqse replied:
I will remove it but why? You directly see if everything worked as it should.

ChristophGr replied:
These "artworks" just clutter a good log. In a demo-app you can consider printing this stuff to stdout directly.

frenzqse committed:
[OPENENGSB-3062] Remove support for cxf plugin in the .Net bridge. and adapte logging infos

frenzqse committed:
[OPENENGSB-3063] Add a readonly Proprey, which stand for constands in the .Net Bridge

frenzqse committed:
[OPENENGSB-3064] Rewrite Code in the .Net bridge
[OPENENGSB-2514] Converting a received exception from openEngSB to a C# exception

frenzqse committed:
[OPENENGSB-3066] Rrewrite code of the factorys in the .Net bridge
[OPENENGSB-3067] Change factory behaviour in the .Net bridge

frenzqse committed:
[OPENENGSB-3070] Remove NPanday-setting location and add csc- and wsdl- location in the wsdl-to-dll plugin
